### PR TITLE
MAM5 aerosol feature for v3 (reworked)

### PR DESCRIPTION
**Pull request by keziming**
This branch is to replace #5703 . 

The new stratospheric sulfate aerosol feature (MAM5) is to represent the stratospheric sulfate aerosol with a minimum increase of the computational cost by adding a new Stratospheric Coarse Mode on top of the default four-mode version of Modal Aerosol Model (MAM4) in E3SMv2. Because this implementation applies stratospheric treatments to a separate aerosol mode, it avoids disrupting the model simulation of tropospheric coarse mode dust and sea salt (Visioni et al., 2021). With this new feature, the E3SM simulated stratospheric sulfate burden and aerosol optical depth (AOD) agree well with satellite observations, e.g., after Mt. Pinatubo eruption in June 1991. More details can be found at [https://acme-climate.atlassian.net/wiki/spaces/NGDAP/pages/3441164365/New+Aerosol+Feature+3+Stratospheric+Sulfate+for+E3SMv3]

The long-term tests of MAM5 are documented in this page.
https://acme-climate.atlassian.net/wiki/spaces/NGDAP/pages/3795222529/MAM5+features

The E3SM tests are documented here
https://acme-climate.atlassian.net/wiki/spaces/ATMOS/pages/3807051811/MAM5+Integration+Tests

### CODE REVIEW

File: components/eam/src/chemistry/modal_aero/modal_aero_data.F90 [hunk 1]
```diff
@@ -97,6 +99,13 @@ module modal_aero_data
          'fine_dust       ', &
          'coarse_seasalt  ', &
          'coarse_dust     '/)
+#elif ( defined MODAL_AERO_5MODE)
+    character(len=*), parameter :: modename_amode(ntot_amode) = (/ &
+         'accum           ', &
+         'aitken          ', &
+         'coarse          ', &
+         'primary_carbon  ', &
+         'strat_coarse  '/)
```
Comment (wlin7) [inline components/eam/src/chemistry/modal_aero/modal_aero_data.F90:108]: The last element need to add additional space to have the same length as the other fields, otherwise gnu compiler would fail (string array constructor).

File: components/eam/src/chemistry/mozart/mo_gas_phase_chemdr.F90 [hunk 2]
```diff
@@ -1258,16 +1294,26 @@ subroutine gas_phase_chemdr(lchnk, ncol, imozart, q, &
...
+                                invariants(:,:,indexm), invariants, del_h2so4_gasprod,        &
+                                vmr0, vmr, pbuf, troplev)
+#else
+       ! this option is left for non modal aerosol
+       call aero_model_gasaerexch( imozart-1, ncol, lchnk, delt, latndx, lonndx, reaction_rates, &
+                                tfld, pmid, pdel, mbar, relhum,                               &
+                                zm,  qh2o, cwat, cldfr, ncldwtr,                              &
+                                invariants(:,:,indexm), invariants, del_h2so4_gasprod,        &
+                                vmr0, vmr, pbuf)
+#endif
```
Comment (mingxuanwupnnl) [inline components/eam/src/chemistry/mozart/mo_gas_phase_chemdr.F90:1315]: I know it works, It would be better to use || like other places instead of |?

File: components/eam/src/chemistry/utils/modal_aero_deposition.F90 [hunk 3]
```diff
@@ -440,8 +441,9 @@ subroutine set_srf_drydep(aerdepdryis, aerdepdrycw, cam_out)
                             0.289_r8*(aerdepdryis(i,idx_dst3)+aerdepdrycw(i,idx_dst3)))
 
 #endif
-
-#if( (defined MODAL_AERO_4MODE) || (defined MODAL_AERO_4MODE_MOM) )
+!kzm ++
+#if( (defined MODAL_AERO_4MODE) || (defined MODAL_AERO_4MODE_MOM) || (defined MODAL_AERO_5MODE))
+!kzm --
```
Comment (mingxuanwupnnl) [inline components/eam/src/chemistry/utils/modal_aero_deposition.F90:446]: remove comments for name

File: components/eam/src/chemistry/utils/modal_aero_deposition.F90 [hunk 4]
```diff
@@ -226,8 +226,9 @@ subroutine set_srf_wetdep(aerdepwetis, aerdepwetcw, cam_out)
                              0.289_r8*(aerdepwetis(i,idx_dst3)+aerdepwetcw(i,idx_dst3)))
 
 #endif
-
-#if( (defined MODAL_AERO_4MODE) || (defined MODAL_AERO_4MODE_MOM) )
+!kzm ++
+#if( (defined MODAL_AERO_4MODE) || (defined MODAL_AERO_4MODE_MOM) || (defined MODAL_AERO_5MODE))
+!kzm --
```
Comment (mingxuanwupnnl) [inline components/eam/src/chemistry/utils/modal_aero_deposition.F90:231]: remove pure name comments

File: components/eam/src/physics/cam/aer_rad_props.F90 [hunk 5]
```diff
@@ -77,7 +79,21 @@ subroutine aer_rad_props_init()
...
+   !call add_hist_coord('nswbands',    nswbands,    'NSWBANDS')   
+   !call addfld ('ext_sao_lw',(/ 'lev     ', 'nlwbands' /),    'A','1/m',&
+   !     'Aerosol LW radiation properties comparable to prescribed input ext_earth', flag_xyfill=.true.) 
+   !call addfld ('ext_sao_sw',(/ 'lev     ', 'nswbands' /),    'A','1/m',&
+   !     'Aerosol SW radiation properties comparable to prescribed input ext_sun', flag_xyfill=.true.) 
+   !call addfld ('ssa_sao',(/ 'lev     ', 'nswbands' /),    'A','1/m',&
+   !     'Aerosol SW radiation properties comparable to prescribed input omega_sun', flag_xyfill=.true.) 
+   !call addfld ('af_sao',(/ 'lev     ', 'nswbands' /),    'A','1/m',&
+   !     'Aerosol SW radiation properties comparable to prescribed input g_sun', flag_xyfill=.true.) 
+   !endif
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/aer_rad_props.F90:96]: If they are just some temporary output. Since those are commented out, maybe we can remove them. If it is some diagnostics we want to add, we can either uncomment them or make a separate PR.

File: components/eam/src/physics/cam/aer_rad_props.F90 [hunk 6]
```diff
@@ -197,6 +213,12 @@ subroutine aer_rad_props_sw(list_idx, dt, state, pbuf,  nnite, idxnite, is_cmip6
    real(r8), pointer :: ext_cmip6_sw(:,:,:)
    real(r8) :: ext_cmip6_sw_inv_m(pcols,pver,nswbands)! short wave extinction in the units of 1/m
 
+   ! for Strat. AOD output
+   !if (is_output_interactive_volc) then 
+   !real(r8) :: ext_sao_sw(pcols,0:pver,nswbands) !
+   !real(r8) :: ssa_sao   (pcols,0:pver,nswbands) !
+   !real(r8) :: af_sao    (pcols,0:pver,nswbands) !
+   !endif
    !-----------------------------------------------------------------------------
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/aer_rad_props.F90:222]: same thing here. Either remove it for cleaner merge or uncomment them.

File: components/eam/src/physics/cam/aer_rad_props.F90 [hunk 7]
```diff
@@ -491,7 +531,17 @@ subroutine aer_rad_props_lw(is_cmip6_volc, list_idx, dt, state, pbuf,  odap_aer,
...
+   !  ext_sao_lw(:,:,:) = 0.0_r8
+     !prepare strat. aerosol sw optic properties for output: 
+   !  do ipver = 1 , pver
+   !      do icol = 1, ncol
+   !         lyr_thk = state%zi(icol,ipver) - state%zi(icol,ipver+1)
+   !         ext_sao_lw(icol,ipver,:) = odap_aer(icol,ipver,:)/lyr_thk ! unit m
+   !      enddo
+   !  enddo    
+   !  call outfld('ext_sao_lw',ext_sao_lw(:,:,:), pcols, lchnk)
+   !endif
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/aer_rad_props.F90:544]: same thing here

File: components/eam/src/physics/cam/modal_aer_opt.F90 [hunk 8]
```diff
@@ -553,11 +586,25 @@ subroutine modal_aero_sw(list_idx, dt, state, pbuf, nnite, idxnite, is_cmip6_vol
...
+   !ihuge = huge(ihuge)
+   !trop_level(:) = ihuge
+   !call tropopause_find(state, trop_level)!
+   !Quit if tropopause is not found
+   !if (any(trop_level(1:ncol) == -1)) then
+   !   do icol = 1, ncol
+   !      write(iulog,*)'tropopause level,lchnk,column:',trop_level(icol),lchnk,icol
+   !   enddo
+   !   call endrun('aer_rad_props_lw: tropopause not found')
+   !endif
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/modal_aer_opt.F90:607]: same issue mentioned before, add them or remove them

File: components/eam/src/physics/cam/nucleate_ice_cam.F90 [hunk 9]
```diff
@@ -344,6 +356,39 @@ subroutine nucleate_ice_cam_init(mincld_in, bulk_scale_in)
...
+               accum_bc_idx = n     ! find bc
+            case ('dust')
+               accum_dst_idx = n    ! find dust
+            case ('seasalt')
+               accum_ncl_idx = n    ! find dust 
+            case ('m-organic')
+               accum_mom_idx = n    ! find dust   
+            end select
+         end do
+      endif
```
Comment (mingxuanwupnnl) [inline components/eam/src/physics/cam/nucleate_ice_cam.F90:391]: Do we need those accum variables?

File: components/eam/bld/namelist_files/use_cases/20TR_eam_chemUCI-Linoz-mam5.xml [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+<airpl_emis_file></airpl_emis_file>   <!-- need to be empty, but if specifying empty here, the value would be root of input_data_path -->
+
+<!-- Prescribed oxidants for aerosol chemistry.   Ozone is from CMIP6 input4MIPS file -->
+<tracer_cnst_type    >CYCLICAL</tracer_cnst_type>
+<tracer_cnst_cycle_yr>1995</tracer_cnst_cycle_yr>
+<tracer_cnst_file    >ch4_oxid_1.9x2.5_L26_1990-1999clim.c090804.nc</tracer_cnst_file>
+<tracer_cnst_filelist>''</tracer_cnst_filelist>
+<tracer_cnst_datapath>atm/cam/chem/methane</tracer_cnst_datapath>
+<tracer_cnst_specifier>'CH4','prsd_NO3:NO3', 'prsd_OH:OH'</tracer_cnst_specifier>
```
Comment (mingxuanwupnnl) [inline components/eam/bld/namelist_files/use_cases/20TR_eam_chemUCI-Linoz-mam5.xml:74]: Just want to note here, it would be nice in the future before v3 freeze we switch to use historical series of oxidants for NO3, OH. No need to change for now

File: components/eam/cime_config/config_component.xml [hunk 11]
```diff
@@ -53,6 +53,7 @@
       <value compset="_EAM%AR5sf"      >&eam_phys_defaults; -chem superfast_mam4_resus_mom_soag -rain_evap_to_coarse_aero</value>
       <value compset="_EAM%CHEMUCI-LINOZV2"  >-clubb_sgs -microphys mg2 -chem superfast_mam4_resus_mom_soag -rain_evap_to_coarse_aero -nlev 72 -usr_mech_infile $SRCROOT/components/eam/chem_proc/inputs/pp_chemUCI_mam4_resus_mom_soag_tag.in</value>
       <value compset="_EAM%CHEMUCI-LINOZV3"  >-clubb_sgs -microphys mg2 -chem superfast_mam4_resus_mom_soag -rain_evap_to_coarse_aero -nlev 72 -usr_mech_infile $SRCROOT/components/eam/chem_proc/inputs/pp_chemUCI_linozv3_mam4_resus_mom_soag_tag.in</value>
+      <value compset="_EAM%CHEMUCI-LINOZV3-MAM5"  >-clubb_sgs -microphys mg2 -chem superfast_mam5_resus_mom_soag -rain_evap_to_coarse_aero -nlev 72 -usr_mech_infile $SRCROOT/components/eam/chem_proc/inputs/pp_chemUCI_linozv3_mam5_resus_mom_soag_tag_cnst_no3_oh.in</value>
       <value compset="_ELM%[^_]*BC"    >-bc_dep_to_snow_updates</value>
```
Comment (mingxuanwupnnl) [inline components/eam/cime_config/config_component.xml:57]: We will probably have a new compset. Just note that microphys should be p3 in the new ones

File: components/eam/src/chemistry/modal_aero/modal_aero_data.F90 (hunk 1 above)
Comment (keziming) [reply components/eam/src/chemistry/modal_aero/modal_aero_data.F90]: @wlin7 thank you for finding it. I pushed the commit to remote

File: components/eam/src/chemistry/modal_aero/aero_model.F90 [hunk 12]
```diff
@@ -841,6 +841,46 @@ subroutine aero_model_init( pbuf2d, species_class, iflagaa )
...
+    index_tot_mass(2,3) = get_spc_ndx('ncl_a2')
+    index_tot_mass(2,4) = get_spc_ndx('mom_a2')
+    index_chm_mass(2,1) = get_spc_ndx('so4_a2')
+    index_chm_mass(2,2) = get_spc_ndx('soa_a2')
+    ! coarse mode
+    !
+    index_tot_mass(3,1) = get_spc_ndx('dst_a3')
+    index_tot_mass(3,2) = get_spc_ndx('ncl_a3')
+    index_tot_mass(3,3) = get_spc_ndx('so4_a3')
+    index_chm_mass(3,1) = get_spc_ndx('so4_a3')
```
Comment (mingxuanwupnnl) [inline components/eam/src/chemistry/modal_aero/aero_model.F90:871]: I wonder if we should follow accumulation mode here to also include bc, pom, soa, mom for index_tot_mass. As well as making changes for index_chm_mass. We know that in v2 those does not really affect simulations, as it used prescibed SAD. Now, since we use diagnostic calculation, it may matter.

File: components/eam/src/chemistry/mozart/mo_gas_phase_chemdr.F90 (hunk 2 above)
Comment (keziming) [reply components/eam/src/chemistry/mozart/mo_gas_phase_chemdr.F90:1315]: thanks. corrected

File: components/eam/src/physics/cam/aer_rad_props.F90 (hunk 5 above)
Comment (keziming) [reply components/eam/src/physics/cam/aer_rad_props.F90:96]: this is not necessary output for now. You are right, I can add it later

File: components/eam/src/physics/cam/aer_rad_props.F90 (hunk 6 above)
Comment (keziming) [reply components/eam/src/physics/cam/aer_rad_props.F90:222]: deleted it.

File: components/eam/src/physics/cam/nucleate_ice_cam.F90 (hunk 9 above)
Comment (keziming) [reply components/eam/src/physics/cam/nucleate_ice_cam.F90:391]: I want to keep this part. It is useful to add in better ice nucleation for the future and it doesn't consume much computational resource.

File: components/eam/bld/namelist_files/use_cases/20TR_eam_chemUCI-Linoz-mam5.xml (hunk 10 above)
Comment (keziming) [reply components/eam/bld/namelist_files/use_cases/20TR_eam_chemUCI-Linoz-mam5.xml:74]: Thanks for the note

File: components/eam/cime_config/config_component.xml (hunk 11 above)
Comment (keziming) [reply components/eam/cime_config/config_component.xml:57]: this compset if for the MAM5 PR test purpose.

File: components/eam/src/chemistry/modal_aero/aero_model.F90 (hunk 12 above)
Comment (keziming) [reply components/eam/src/chemistry/modal_aero/aero_model.F90:871]: I agree with you. But in chemUCI we don't use aerosol heterogeneous reactions yet. It should have no impact for now. Anyway, I changed the code and we may use it in the near future.

**Comment by keziming** [comment]
> I saw that there are still two files having conflicts. Have we already resolved them? Or we can ignore them?

This will be solve when merge it. They are very minor conflicts. Last time I solved the conflicts on line, github merged master into my code base.

**Comment by mingxuanwupnnl** [comment]
@keziming Thanks for the prompt response. I don't think the changes I suggested would change answer, but please revert them if you see any non BFB issues, like the changes for index_tot_mass in aero_model.F90. I think this PR is ready to go.

**Comment by rljacob** [comment]
I changed the base branch so this is merged to v3atm/eam/master_MAM5_wetaero_chemdyg.  @keziming please update the description.
